cirq-core/cirq/qis/clifford_tableau.py

```diff
@@ -176,6 +179,76 @@ def _str_full_(self) -> str:
 
         return string
 
+    def then(self, second: 'CliffordTableau') -> 'CliffordTableau':
+        """Returns a composed CliffordTableau of this tableau and the second tableau.
+
+        Then composed tableau is equal to (up to global phase) the composed
+        unitary operation of the two tableaux, i.e. equivalent to applying the unitary
+        operation of this CliffordTableau then applying the second one.
+
+        Args:
+            second: The second CliffordTableau to compose with.
+
+        Returns:
+            The composed CliffordTableau.
+
+        Raises:
+            TypeError: If the type of second is not CliffordTableau.
+            ValueError: If the number of qubits in the second tableau mismatch with
+                this tableau.
+        """
+        if not isinstance(second, CliffordTableau):
+            raise TypeError("The type for second tableau must be the CliffordTableau type")
+        if self.n != second.n:
+            raise ValueError(
+                f"Mismatched number of qubits of two tableaux: {self.n} vs {second.n}."
+            )
+
+        # Convert the underlying data type from bool to int for easier numerical computation.
+        m1 = self.matrix().astype(int)
+        m2 = second.matrix().astype(int)
+
+        # The following computation is based on Theorem 36 in
+        # https://arxiv.org/pdf/2009.03218.pdf.
+        # Any pauli string (one stabilizer) in Clifford Tableau should be able to be expressed as
+        #    (1i)^p (-1)^s X^(mx) Z^(mz)
+        # where p and s are binary scalar and mx and mz are binary vectors.
+        num_ys1 = np.sum(m1[:, : self.n] * m1[:, self.n :], axis=1)
+        num_ys2 = np.sum(m2[:, : self.n] * m2[:, self.n :], axis=1)
+
+        p1 = np.mod(num_ys1, 2)
+        p2 = np.mod(num_ys2, 2)
+
+        s1 = self.rs.astype(int) + np.mod(num_ys1, 4) // 2
```

Comment:
Explain the conversion between `s` and `r` (maybe using the YYY example we talked through).

Reply:
Done